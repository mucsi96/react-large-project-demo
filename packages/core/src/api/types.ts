export type ApiMethod = 'GET' | 'POST' | 'PUT' | 'DELETE';

export type CallApiOptions = {
  href: string;
  queryParams?: Record<string, string>;
  method?: ApiMethod;
  headers?: HeadersInit;
  signal?: AbortSignal | null;
  body?: unknown;
  cache?: boolean;
};

export type ApiCaller<ResponseBody = unknown> = (
  options: CallApiOptions
) => Promise<ResponseBody>;

export type ApiErrorResponse = {
  error?: {
    message?: string;
  };
};

export type ApiError = {
  message: string;
  response?: ApiErrorResponse;
  status?: number;
};

export type ApiState<FetchArgs extends unknown[], ResponseBody> = {
  data?: ResponseBody;
  error?: ApiError;
  isLoading: boolean;
  fetchArgs?: FetchArgs;
};
