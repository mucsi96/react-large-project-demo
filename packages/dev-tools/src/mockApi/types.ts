export type MockMethod = "GET" | "POST" | "DELETE" | "PUT";

export type MockRequest = {
  url: string;
  method: MockMethod;
  headers: Record<string, string | string[]>;
  body: object;
  params: Record<string, string | string[]>;
  query: Record<string, string | string[]>;
};

export type MockResponse = {};

export type Mock = {
  path: string;
  method?: MockMethod;
  callback: (
    req: MockRequest,
    res: MockResponse
  ) => Promise<object | undefined> | object | undefined;
};

export type MockWithRegexp = Mock & {
  regexp: RegExp;
  keys: string[];
};
