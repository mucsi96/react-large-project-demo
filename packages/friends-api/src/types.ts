export type Link = {
  href: string;
  method?: string;
};

export type Friend = {
  id: string;
  firstName: string;
  lastName: string;
  image: string;
};

export type FriendsResponse = {
  _embedded: Friend[];
  _links: {
    next?: Link;
  };
};
