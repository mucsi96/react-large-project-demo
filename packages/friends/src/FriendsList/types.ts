import { Friend, FriendsResponse } from 'friends-api';

type LoadFriendsAction = {
  type: 'LOAD_FRIENDS';
  payload: FriendsResponse;
};

export type FriendsAction = LoadFriendsAction;

export type FriendsState = {
  friends: Friend[];
};
