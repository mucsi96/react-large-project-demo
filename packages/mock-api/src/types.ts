export type MockMethod = 'GET' | 'POST' | 'DELETE' | 'PUT';

export type MockRequest = {
  url: string;
  method: MockMethod;
  headers: Record<string, string | string[]>;
  body?: unknown;
  params: Record<string, string>;
  query: Record<string, string | string[]>;
};

export type MockResponse = {
  status: (statusCode: number) => void;
  delay: (delayMs: number) => void;
  mockError: (enable: boolean) => void;
  mockHTML: (enable: boolean) => void;
};

export type Mock = {
  path: string;
  method?: MockMethod;
  callback: (
    req: MockRequest,
    res: MockResponse
  ) => Promise<unknown | undefined> | unknown | undefined;
};

export type MockWithRegexp = Mock & {
  regexp: RegExp;
  keys: string[];
};
