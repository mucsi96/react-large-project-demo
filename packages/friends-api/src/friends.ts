import { fetchJSON } from 'core';
import { FriendsResponse, Link } from './types';

export function getFriends(
  link: Link = { href: '/api/friends' }
): Promise<FriendsResponse> {
  return fetchJSON(link.href);
}
