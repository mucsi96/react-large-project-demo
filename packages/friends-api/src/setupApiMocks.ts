import {
  getMockSwitch,
  registerApiMocks,
  MockRequest,
  MockResponse,
} from 'dev-tools';
import mockFriends from './mockFriends';
import { FriendsResponse } from './types';

export function setupApiMocks(): void {
  registerApiMocks([
    {
      path: '/api/friends',
      callback: getFriends,
    },
  ]);
}

function getFriends({ query }: MockRequest, response: MockResponse) {
  switch (getMockSwitch('friends')) {
    case 'processingFailure':
      return response.mockError(true);
    default:
      const from = parseInt((query.from as string) ?? '0');
      const to = from + 5;
      return {
        _embedded: mockFriends.slice(from, to),
        _links: {
          next: mockFriends[to]
            ? {
                href: `/api/friends?from=${to}`,
              }
            : undefined,
        },
      } as FriendsResponse;
  }
}
