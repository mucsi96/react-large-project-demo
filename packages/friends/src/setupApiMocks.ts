import { setupApiMocks as setupFriendsApiMocks } from 'friends-api';

export function setupApiMocks(): void {
  setupFriendsApiMocks();
}
