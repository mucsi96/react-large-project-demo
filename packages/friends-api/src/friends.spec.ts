import { getFriends } from './friends';
import { setupApiMocks } from './setupApiMocks';
import { FriendsResponse } from './types';

setupApiMocks();

describe('friends', () => {
  describe('getFriends', () => {
    test('fetches friends', async () => {
      const friends = await getFriends();
      expect(reduceImages(friends)).toMatchSnapshot();
    });

    it('fetches the next page of friends', async () => {
      const friends1 = await getFriends();
      const friends2 = await getFriends(friends1._links.next);
      expect(reduceImages(friends2)).toMatchSnapshot();
    });

    it('returns no next link if there are no more friends', async () => {
      const friends1 = await getFriends();
      const friends2 = await getFriends(friends1._links.next);
      const friends3 = await getFriends(friends2._links.next);
      expect(friends3._links.next).toBeUndefined();
    });
  });
});

function reduceImages(friends: FriendsResponse): FriendsResponse {
  return {
    ...friends,
    _embedded: friends._embedded.map((friend) => ({
      ...friend,
      image: friend.image.substring(friend.image.length - 100),
    })),
  };
}
