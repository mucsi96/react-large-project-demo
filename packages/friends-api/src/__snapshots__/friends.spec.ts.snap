// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`friends getFriends fetches friends 1`] = `
Object {
  "_embedded": Array [
    Object {
      "firstName": "Alyson",
      "id": "alyson",
      "image": "wNMM9flXbj/x7NR6y5ku5Jm5eRy5PuTmiirk37OPzM6atVl6L9StE0gf7/atXTr6WK42g9BRRWDOpGy1w0sal+frRRRWYmkf/9k=",
      "lastName": "Donnelly",
    },
    Object {
      "firstName": "Carlee",
      "id": "carlee",
      "image": "c3czsJTjpmuX8UzHyHJoooA8S1XXPMvRbSW6kxyN8+eSCMVJb3UcEcAkWR03naobAGetFFAHfadZWUcKmONwGGe3eiiigD//2Q==",
      "lastName": "Kreiger",
    },
    Object {
      "firstName": "Enrico",
      "id": "enrico",
      "image": "QueHjW3USKipufaxyB8ze9cN41u2u9UNjyI4MM3+0xGfyAP6miiljpNUnYvARTrK5jpaqopTAooorwbn0NkMaIZoooqkRY//2Q==",
      "lastName": "Pouros",
    },
    Object {
      "firstName": "Patricia",
      "id": "patricia",
      "image": "XkUaZSIMVBOcc0UV1YZ6tHBjErXPrT4datFc2DadBbLbx2KKkYTps7D68V1znIINFFclZWk0jSk7xTZnz8NxRRRXnS3O6Ox//9k=",
      "lastName": "Barrows",
    },
    Object {
      "firstName": "Steven",
      "id": "bergstrom",
      "image": "+IVAcRQr92JPQe/HJ70UUAUEkXKhQdp5+p96Gb5SQMFTng9aKKAIInL3B2jsTyaspM6qN2OOp9RxRRQBOHYncmAenIooooA//9k=",
      "lastName": "Bergstrom",
    },
  ],
  "_links": Object {
    "next": Object {
      "href": "/api/friends?from=5",
    },
  },
}
`;

exports[`friends getFriends fetches the next page of friends 1`] = `
Object {
  "_embedded": Array [
    Object {
      "firstName": "Carol",
      "id": "carol",
      "image": "TLBRKTkHkY9PeiisurOxbI9t+HQNvr17aYUhVaQMB2ZkYA/TJrrtdhB1KbnrjP5CiivMqblSdpmFe2bbV2OBz3oooqEaJn//2Q==",
      "lastName": "Fahey",
    },
    Object {
      "firstName": "Kevin",
      "id": "kevin",
      "image": "biiilIEfbXhGcyeFdIeUl3a0iLE9zsGai8R2kNzYzFVMcoU4dTgiiivOkkehHY8LvLlNWd7HVI/OltnJD9c9qKKKzsiVqtT/2Q==",
      "lastName": "Flatley",
    },
    Object {
      "firstName": "Michelle",
      "id": "michelle",
      "image": "NT/6H+lfVn7PsS2HhR7VQCBO5BHpuOM0UUou8gklynqm7jHvTYWzF+JH60UVsc4+gnFFFAETHiqdyOKKKQzFuQBIfeiiikM//9k=",
      "lastName": "Gibson",
    },
    Object {
      "firstName": "Larry",
      "id": "larry",
      "image": "a1OnP6s4KMIuye5kpzj0NQXLFyFPQUUV7Utj5aHxD5VwY19qRQMkUUUuodBVba4I78GrBj+Y4PWiiqRMgi5Y56elFFFNbEPc/9k=",
      "lastName": "Goodwin",
    },
    Object {
      "firstName": "Sarah",
      "id": "sarah",
      "image": "te6HKXadbm2bl0YYJ9x71fe4R7iFzvAkwAR1wexoopT96Km92KEFCcqcdlYxtbRoJ3idstGwG4DqCMiiiiuum/dR5VWK52f/2Q==",
      "lastName": "Graham",
    },
  ],
  "_links": Object {
    "next": Object {
      "href": "/api/friends?from=10",
    },
  },
}
`;
